Provide script that allows to get all tasks with a certain property

lembergerth opened the pull request:
In case people do not use the benchexec infrastructure,
it is difficult to get the full set of benchmark files
with a certain property.

This script makes that task easier:
It gets a property name and, optionally, the benchmark directory,
and prints all tasks with the given property to stdout.
At the moment, properties must be given without a directory prefix to
reliably work. This way, it is not necessary to look at each single task
definition file separately, but it may be imprecise in case
two different properties with the same basename exist in different
directories.

Example usage:
```
./get_tasks_for_property.sh coverage-branches.prp c/
```

PhilippWendler commented:
We should not provide quick-and-dirty hacks like this as official examples in the repo. If we add such examples, they should use the proper way of parsing the Yaml. There is no reason not do so, because this is trivially done with `yq`.

And probably it makes more sense to print the name of the task-definition file instead of the content of key `input_files` (which could even be a list), because the task-definition file is much more useful for further processing (for example, the input files can be extracted from it, but other way round is not possible):

PhilippWendler commented:
Example using string matching:
```
yq -e  '.properties | map (.property_file | contains("/unreach-call.prp")) | any' c/loops/terminator_03-1.yml
```

Even better example that actually checks that the property file is really the same and resolves symlinks etc. just like BenchExec:
```
for task in loops/*.yml; do
  for prp in $(yq  -r '.properties[].property_file' "$task" ); do
    if [ properties/termination.prp -ef "$(dirname "$task")/$prp" ]; then
      echo "$task"
      break
    fi
  done
done
```

lembergerth commented:
Thanks for the examples! I incorporated the second, adjusted the yq-parsing to work with yml-files that are no task definition, and actually return the yml file now.

I understand that this is necessary to make the script work with multiple input files,
but think this makes it a lot less usable: The actual use-case is that people that are inexperienced
with the shell or yml-parsing can get the C or Java files that were used in a certain category of SV-COMP or Test-Comp.

PhilippWendler commented:
I would remove the `grep` hack completely. This hack can silently break in some cases, leading to potentially wrong benchmark results or invalid data analysis or whatever people use this example for. Performance should not be prioritized over correctness, especially not in example code which we expect to be copied and reused.

If you think that people often want to get the set of input files, why not provide another example or some documentation how to do this? It is as easy as running another single-line `yq` command, though I would suggest anyone who wants to more with these tasks to just write a few lines of Python (which is probably even less effort than getting a properly working shell script).

lembergerth commented:
> I would remove the grep hack completely. This hack can silently break in some cases, leading to potentially wrong benchmark results or invalid data analysis or whatever people use this example for.

Why do you consider this a "hack"? Can you provide an example for when this can break?
This provides a significant speedup, at least for Test-Comp tasks.

> It is as easy as running another single-line yq command, though I would suggest anyone who wants to more with these tasks to just write a few lines of Python

So you expect people to either learn jq or the Python yaml module.

PhilippWendler commented:
> > I would remove the grep hack completely. This hack can silently break in some cases, leading to potentially wrong benchmark results or invalid data analysis or whatever people use this example for.
> 
> Why do you consider this a "hack"? Can you provide an example for when this can break?

You yourself provide an example for when this can break:
https://github.com/sosy-lab/sv-benchmarks/blob/2b8090c7d5d2c498dcb957fde26e0cab0541672b/get_tasks_for_property.sh#L27-L28

>  This provides a significant speedup, at least for Test-Comp tasks.

How often do you run this script that this is a real problem? Why not temporarily store the results of the script if you need them more than once?
And for your own use case, you can of course still make certain assumptions and speedup things, but please do not use this as an example. People will assume that they can just copy that code and it works.

> So you expect people to either learn jq or the Python yaml module.

Or use their favorite programming language. Or use the provided example.

lembergerth commented:
> You yourself provide an example for when this can break:

An example that doesn't occur in sv-benchmarks and which can be quickly changed once it occurs (thanks to the comment).
But ok, I don't want to continue arguing; I've removed that optimization.

> How often do you run this script that this is a real problem? Why not temporarily store the results of the script if you need them more than once?

Every time after a pull, so that my file with these tasks is up-to-date.

> People will assume that they can just copy that code and it works.

But they can.

dbeyer commented:
@PhilippWendler Could you please review this again?

dbeyer commented:
@lembergerth Please get a review for your changes. ;-)

lembergerth commented:
Yes, sorry, and thanks for the hints!
Do you also want me to squash merge this branch before merging?
(btw, someone will have to merge the branch for me, I'm not allowed to do that)
